File: src/main/java/de/fluxparticle/animation/signal/SignalPath.java
package de.fluxparticle.animation.signal;

import java.util.function.Function;

public abstract class SignalPath<T> implements Function<Double, T> {

    public static abstract class Builder<T> {

        protected T start;

        protected T end;

        public Builder<T> start(T start) {
            this.start = start;
            return this;
        }

        public Builder<T> end(T end) {
            this.end = end;
            return this;
        }

        public abstract SignalPath<T> build();

    }

    protected final T start;

    protected final T end;

	public SignalPath(T start, T end) {
		this.start = start;
		this.end = end;
	}

}

File: src/main/kotlin/de/fluxparticle/animation/signal/SignalPath.kt
package de.fluxparticle.animation.signal

import java.util.function.Function

abstract class SignalPath<T>(protected val start: T, protected val end: T) : Function<Double, T> {

    abstract class Builder<T : Any> {

        protected lateinit var start: T

        protected lateinit var end: T

        fun start(start: T): Builder<T> {
            this.start = start
            return this
        }

        fun end(end: T): Builder<T> {
            this.end = end
            return this
        }

        abstract fun build(): SignalPath<T>

    }

}
